### DISCUSSION TITLE
WebSockets issue

### DISCUSSION THREAD

**Discussion by shawn-digitalpoint** [btcpayserver/btcpayserver#2507] [Technical Support] [answered]
I have BTCPay Server running fine/as expected, however the WebSockets portion of it doesn't work.  It's not a catastrophic issue, but it would be nice if it works because sometimes I want to be able to sign an outgoing transaction with multiple outputs with a hardware wallet.

My Nginx config is proxying other WebSocket connections without issue, but I decided to test it without Nginx (to see if it was an issue with the Nginx config or btcpayserver).

This is what I tested:

---
## Normal HTTP connection (works):
`docker exec -it generated_btcpayserver_1 curl -i -N http://localhost:49392

HTTP/1.1 302 Found
Date: Wed, 05 May 2021 16:25:20 GMT
Server: Kestrel
Content-Length: 0
Location: http://localhost:49392/login?ReturnUrl=%2F
Referrer-Policy: same-origin
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: DENY`

---
## Normal HTTPS connection (fails):
`docker exec -it generated_btcpayserver_1 curl -i -N https://localhost:49392

curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number`

---
## Known Working WebSocket connection (works):
`docker exec -it generated_btcpayserver_1 curl -i -N -H "Connection: Upgrade" -H "Upgrade: websocket" -H "Host: echo.websocket.org" -H "Origin: https://www.websocket.org" https://echo.websocket.org

HTTP/1.1 101 Web Socket Protocol Handshake
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: content-type
Access-Control-Allow-Headers: authorization
Access-Control-Allow-Headers: x-websocket-extensions
Access-Control-Allow-Headers: x-websocket-version
Access-Control-Allow-Headers: x-websocket-protocol
Access-Control-Allow-Origin: https://www.websocket.org
Connection: Upgrade
Date: Wed, 05 May 2021 16:29:12 GMT
Server: Kaazing Gateway
Upgrade: WebSocket
WebSocket-Location: wss://echo.websocket.org/
WebSocket-Origin: https://www.websocket.org`

---
## BTCPay Server HTTP WebSocket connection (fails)
`docker exec -it generated_btcpayserver_1 curl -i -N -H "Connection: Upgrade" -H "Upgrade: websocket" -H 'Host: [redacted]' -H 'Origin: [redacted]' -H 'Cookie: [authentication cookies redacted]' http://localhost:49392/Notifications/SubscribeUpdates

HTTP/1.1 400 Bad Request
Connection: close
Date: Wed, 05 May 2021 16:32:05 GMT
Server: Kestrel
Content-Length: 0
Referrer-Policy: same-origin
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: DENY`

---
## BTCPay Server HTTPS WebSocket connection (fails)
`docker exec -it generated_btcpayserver_1 curl -i -N -H "Connection: Upgrade" -H "Upgrade: websocket" -H 'Host: [redacted]' -H 'Origin: [redacted]' -H 'Cookie: [authentication cookies redacted]' https://localhost:49392/Notifications/SubscribeUpdates

curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number`

---
## BTCPay Server SSL Handshake (fails)
`docker exec -it generated_btcpayserver_1 openssl s_client -connect localhost:49392 -msg

CONNECTED(00000003)
\>>> ??? [length 0005]
    16 03 01 01 16
\>>> TLS 1.3, Handshake [length 0116], ClientHello
    01 00 01 12 03 03 2f d0 92 fb bc 37 14 0f 29 56
    16 da 3e 62 07 ae e2 0e 07 bf 00 2c e2 ea 0c 52
    c8 06 e1 fa a2 ac 20 6a 8e bf ca 6b e6 92 45 20
    93 7f 95 f5 66 c4 99 2f 8f ed 46 42 c7 42 0b ce
    89 8b 50 70 af f3 53 00 3e 13 02 13 03 13 01 c0
    2c c0 30 00 9f cc a9 cc a8 cc aa c0 2b c0 2f 00
    9e c0 24 c0 28 00 6b c0 23 c0 27 00 67 c0 0a c0
    14 00 39 c0 09 c0 13 00 33 00 9d 00 9c 00 3d 00
    3c 00 35 00 2f 00 ff 01 00 00 8b 00 0b 00 04 03
    00 01 02 00 0a 00 0c 00 0a 00 1d 00 17 00 1e 00
    19 00 18 00 23 00 00 00 16 00 00 00 17 00 00 00
    0d 00 2a 00 28 04 03 05 03 06 03 08 07 08 08 08
    09 08 0a 08 0b 08 04 08 05 08 06 04 01 05 01 06
    01 03 03 03 01 03 02 04 02 05 02 06 02 00 2b 00
    05 04 03 04 03 03 00 2d 00 02 01 01 00 33 00 26
    00 24 00 1d 00 20 a5 98 aa 6e 71 5c 64 7e 08 a7
    b0 4f c7 de 44 95 ff f4 26 0a 61 8a ea 4d 3a 98
    e1 5d 1b 33 36 79
\<<< ??? [length 0005]
    48 54 54 50 2f
139758254797952:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:332:
\---
no peer certificate available
\---
No client certificate CA names sent
\---
SSL handshake has read 5 bytes and written 283 bytes
Verification: OK
\---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
\---`

---
My guess is that the non-WebSocket stuff is able to work normally because Nginx can fallback to proxying HTTP, but for WebSocket endpoints, neither HTTP or HTTPS works.

If you take the 5 bytes from the SSL handshake that btcpayserver responds with (it's supposed to be the SSL version info for the handshake), and decode the hex on those 5 bytes (48 54 54 50 2f), it's "HTTP/".  That looks an awful lot like the first 5 bytes of a non-SSL HTTP connection, and pretty clear why the SSL connections to btcpayserver are failing (it's not trying to negotiate the SSL connection).

I don't know enough about the guts of btcpayserver, but my guess is that it's proxying connections to some internal process and that proxy is ignoring SSL.  That in itself is fine since Nginx can proxy HTTP to HTTPS just fine, however, for insecure WebSocket connections, btcpayserver responds with "HTTP/1.1 400 Bad Request".

So I guess my question is this... how do I get btcpayserver to do SSL handshake properly *or* how do I get btcpayserver to allow insecure WebSocket connections so Nginx can proxy WebSockets?

Again, the only reason I even care about WebSockets working for btcpayserver is so I can sign an outgoing transaction with multiple outputs with a hardware wallet, everything else seems to work fine without WebSockets working.

**Comment by dennisreimann** [comment, marked as answer]
Are you using the docker deployment with a custom nginx (as described [here](https://docs.btcpayserver.org/FAQ/FAQ-Deployment/#can-i-use-an-existing-nginx-server-as-a-reverse-proxy-with-ssl-termination)) or the nginx from the docker setup?

Either way, using the docker deployment the nginx is supposed to do the SSL termination. The `generated_btcpayserver_1` only exposes port 80, hence only does http.

**Reply by shawn-digitalpoint** [reply]
Tested it both ways.  At first I used the Nginx from the docker setup, but it was causing some issues with our special firewall configuration and it was putting holes in our firewall that we did not want, so I switched it to a normal Nginx instance with the settings in the link you gave so that docker didn't try to be "smart" and punch unwanted holes in the firewall.  BTCPay Server works fine for us in either scenario, with the exception of WebSockets.

So if Nginx is working as intended and proxying insecure HTTP to btcpayserver (and never HTTPS), I think the only issue I'm having is that btcpayserver is giving a "HTTP/1.1 400 Bad Request" HTTP response on only the WebSocket requests (again, testing it directly to make sure Nginx didn't have a misconfiguration).

**Reply by dennisreimann** [reply]
I tried to verify this and can confirm the requests as posted above don't work for me as well. This is due to the lack of the missing [Sec-Websocket headers](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#client_handshake_request): You need to provide the `Sec-WebSocket-Version` and `Sec-WebSocket-Key` headers to make it work.

**Reply by shawn-digitalpoint** [reply]
Thank you very much... all the info you gave me was very much appreciated and helped me down the right path for debugging.

The short version is that it had to do with Nginx ignoring the needed "Connection" and "Upgrade" headers when HTTP/2 was in play.  Turns out the $http_upgrade and $http_connection Nginx variables were empty even though the client had them in the HTTP request.  Force it so only HTTP/1.1 was allowed and nothing changed in the HTTP request other than HTTP/2 changed to HTTP/1.1 in the HTTP request and Nginx magically had those variables not be empty.  My guess is that internally Nginx ignores or drops those two $http_* variables if the connection is HTTP/2.  Maybe it's part of the HTTP/2 spec that they aren't allowed or something and Nginx is doing the proper thing.  Don't really care enough to figure out the "why" at this point.  Just glad I was finally able to figure out with your help.

**Reply by dennisreimann** [reply]
Ok, glad it worked out for you. So this can be considered resolved?

**Reply by shawn-digitalpoint** [reply]
For me, yes... it seems to be.  I didn't go down the road of trying to fix it in the pre-packaged Nginx docker container that comes with btcpayserver (was doing the same thing for me with the docker version as well as "normal" version of Nginx).  Was easier to disable HTTP/2 on the normal version for me.  But yes... everything good for me.  :)
